packages/backend/server/src/plugins/copilot/context/resolver.ts

```diff
@@ -708,27 +708,43 @@ export class CopilotContextResolver {
       return [];
     }
 
-    const session = await this.context.get(context.id);
-    await this.ac
-      .user(user.id)
-      .workspace(session.workspaceId)
-      .allowLocal()
-      .assert('Workspace.Copilot');
-    const allowEmbedding = await this.models.workspace.allowEmbedding(
-      session.workspaceId
-    );
-    if (!allowEmbedding) {
-      return [];
-    }
-
     try {
-      return await session.matchWorkspaceChunks(
+      const session = await this.context.get(context.id);
+      await this.ac
+        .user(user.id)
+        .workspace(session.workspaceId)
+        .allowLocal()
+        .assert('Workspace.Copilot');
+      const allowEmbedding = await this.models.workspace.allowEmbedding(
+        session.workspaceId
+      );
+      if (!allowEmbedding) {
+        return [];
+      }
+
+      const chunks = await session.matchWorkspaceChunks(
         content,
         limit,
         this.getSignal(ctx.req),
         threshold
       );
+      const docsMap = await Promise.all(
+        chunks.map(c =>
+          this.ac
+            .user(user.id)
+            .workspace(session.workspaceId)
+            .doc(c.docId)
```

Comment:
Add an interface to `ac` for batch checking whether `docId` has the same permission? This way, the database can be queried only once instead of N times

Reply:
prefer to implemen batch check in the new PR because a new access controller needs to be implemented

let's first let the permission check the behavior right